File: NinePSharp.Messages.FSharp/messages/Rlock.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Rlock =
    { Size: uint32
      Tag: uint16
      Status: byte }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Rlock
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Rlock span &offset
            span.[NinePConstants.HeaderSize] <- this.Status

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let status = data.[NinePConstants.HeaderSize]
        { Size = size; Tag = tag; Status = status }

File: NinePSharp/messages/Rlock.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Rlock : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Rlock;
    public ushort Tag { get; }
    public byte Status { get; }

    public Rlock(uint size, ushort tag, byte status)
    {
        Size = size;
        Tag = tag;
        Status = status;
    }

    public Rlock(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        Status = data[NinePConstants.HeaderSize];
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Rlock);
        span[NinePConstants.HeaderSize] = Status;
    }
}
